lib/class-wp-theme-json-gutenberg.php

```diff
@@ -207,57 +207,70 @@ class WP_Theme_JSON_Gutenberg {
 	 * - 'value': path to the value in theme.json and block attributes.
 	 */
 	const PROPERTIES_METADATA = array(
-		'background'       => array(
+		'background'           => array(
 			'value' => array( 'color', 'gradient' ),
 		),
-		'background-color' => array(
+		'background-color'     => array(
 			'value' => array( 'color', 'background' ),
 		),
-		'border-radius'    => array(
+		'border-radius'        => array(
 			'value' => array( 'border', 'radius' ),
 		),
-		'border-color'     => array(
+		// This is deliberately separate to the border-radius entry above.
+		// It allows for backwards compatible shorthand e.g. border-radius: 10px
+		// while below provides a means to explicitly set the CSS property name
+		// where the flat array configuration approach builds an incorrect name.
+		'corner-border-radius' => array(
```

Comment:
@aaronrobertshaw Why is it different than `margin`/`padding` For these as well e need to support `margin: 10px` as well, no?

Reply:
I'm not sure I follow your question sorry. Hopefully the following information clarifies things.

For margin/padding, the normal approach takes the key from this metadata array and then adds the individual sides as a suffix to build the CSS property. For example, `margin` with the properties array of `[ 'left', 'right']` would currently create two CSS properties `margin-left` and `margin-right`. 

The border radius CSS properties for individual corners follow a different format than the current approach for margins & paddings. These CSS properties would need the properties array values to be infixed within the normal `border-radius` key. I felt there was less "magic" going on if the CSS property names could actually just be explicitly set if needed. This also helps for [possible new block supports such as grid gaps](https://github.com/WordPress/gutenberg/pull/32571) where the CSS property name is a different format again.

The original `border-radius` entry in the properties metadata array remains for backwards compatibility. It functions as normal (when there isn't per-corner configuration for border radii). The CSS generation here and in global styles, as well as the block support inline styles, check whether the  properties value is an object and if so uses the keys as the CSS property name rather than generating one from a formula.

I'm definitely open to better ideas. This was all that I came up with that would work.

Reply:
> The original border-radius entry in the properties metadata array remains for backwards compatibility. 

Can you clarify how could this entry be used by third-party devs? (just trying to understand the impact).

--- 
Thanks for the explanation, it makes sense but it seems that your solution is the most generic one and I wonder whether we should just use it for padding/margin to stay consistent?

Also cc @nosolosw @jorgefilipecosta as they know more about these APIs...

Reply:
> Can you clarify how could this entry be used by third-party devs? (just trying to understand the impact).

The idea was that the border radius could be specified via the normal `border.radius` property within the theme.json so there would be no change for third-party devs unless I'm missing something.

When the properties metadata is processed into style declarations both configuration entries would have the same value path. A check for whether the value was a string or not allowed the correct css rule to be generated.

> Thanks for the explanation, it makes sense but it seems that your solution is the most generic one and I wonder whether we should just use it for padding/margin to stay consistent?

Ah, I think I'm following now, thanks for the extra questions.

I spent some time this afternoon to try out consolidating the properties metadata entries into a single entry each for border radius, margin and padding. As part of this I've made additional updates so that those single entries can support both shorthand and longhand values e.g. `margin: 10px`, `margin-left: 10px` or `border-top-left-radius: 2px`.

After doing this the existing tests were all passing along with new tests to ensure both shorthand and longhand rule generation worked. It could do with a bit of an extra test before publishing as I'm not certain how the possibility of a shorthand value from the theme.json will work with the Global Styles UI using a box control expecting split values. I'll take a look at this aspect tomorrow.

Will it be OK to push these new changes to this PR, so the configuration updates are all together and with the discussion around why they are required?

Reply:
It looks like we can simplify the existing code in `trunk` if we take one of the ideas proposed in this PR and push it further: make the `PROPERTIES_METADATA` a direct mapping from the CSS property name and the path where that property is stored in `theme.json`.

```php
const PROPERTIES_METADATA = array(
	'background-color'           => array( 'color', 'background' ),
	'background'                 => array( 'color', 'gradient' ),
	'border-color'               => array( 'border', 'color' ),
	'border-radius'              => array( 'border', 'radius' ),
	'border-top-left-radius'     => array( 'border', 'radius', 'topLeft' ),
	'border-top-right-radius'    => array( 'border', 'radius', 'topRight' ),
	'border-bottom-right-radius' => array( 'border', 'radius', 'bottomRight' ),
	'border-bottom-left-radius'  => array( 'border', 'radius', 'bottomLeft' ),
	'border-style'               => array( 'border', 'style' ),
	'border-width'               => array( 'border', 'width' ),
	'color'                      => array( 'color', 'text' ),
	'font-family'                => array( 'typography', 'fontFamily' ),
	'font-size'                  => array( 'typography', 'fontSize' ),
	'font-style'                 => array( 'typography', 'fontStyle' ),
	'font-weight'                => array( 'typography', 'fontWeight' ),
	'letter-spacing'             => array( 'typography', 'letterSpacing' ),
	'line-height'                => array( 'typography', 'lineHeight' ),
	'margin'                     => array( 'spacing', 'margin' ),
	'margin-top'                 => array( 'spacing', 'margin', 'top' ),
	'margin-right'               => array( 'spacing', 'margin', 'right' ),
	'margin-bottom'              => array( 'spacing', 'margin', 'bottom' ),
	'margin-left'                => array( 'spacing', 'margin', 'left' ),
	'padding'                    => array( 'spacing', 'padding' ),
	'padding-top'                => array( 'spacing', 'padding', 'top' ),
	'padding-right'              => array( 'spacing', 'padding', 'right' ),
	'padding-bottom'             => array( 'spacing', 'padding', 'bottom' ),
	'padding-left'               => array( 'spacing', 'padding', 'left' ),
	'text-decoration'            => array( 'typography', 'textDecoration' ),
	'text-transform'             => array( 'typography', 'textTransform' ),
);
```

The other idea this PR introduces is to allow theme authors to either declare a shorthand (`border-radius`) or longhand (`border-top-left-radius`) via `theme.json` (either `border.radius` or `border.radius.topLeft` would work now). I think this is also nice and something `margin` and `padding` would benefit as well.

If it becomes convoluted to do the two things at once we could land first a PR that just changes the implementation but doesn't change behavior, and a second one after that adds support for both long and shorthands. I'm fine reviewing it altogether now that I'm more familiar with these changes ― I also appreciate that UI changes have been already moved to a different PR.

Reply:
Thanks @nosolosw.

I've pushed changes that flatten the `PROPERTIES_METADATA` array and update where that is used accordingly. It definitely simplified some of the code in `class-wp-theme-json-gutenberg.php`. Tests have been updated to cover shorthand and longhand styles as well. 

The `STYLE_PROPERTY` configuration in `packages/blocks/src/api/constants.js` didn't really appear to gain as much if flattened, given its inclusion of the path to check support as well. In the end, I didn't flatten that config, only update the margin and padding support to be consistent with the explicit CSS property naming approach. That itself saved a few conditionals in `style.js` etc.

After pushing the initial changes, I found that I did need to tweak the spacing panel for the Global Styles sidebar so it handled a theme set value that could be either a shorthand value or array representing each side. Nothing too major.

Now the extent of changes required is known, I'm happy to split this PR if it is too much.

The PR description has been updated to reflect the current nature of changes included in this PR.

Reply:
Thanks! It's a lot cleaner now. We can remove the `has_properties` method as well, it's no longer used.